pwiz_tools/Skyline/Model/Proteome/ProteinAssociation.cs

```diff
@@ -561,11 +601,61 @@ public void ApplyParsimonyOptions(bool groupProteins, bool findMinimalProteinLis
+                if (ReferenceEquals(null, x)) return false;
+                if (ReferenceEquals(null, y)) return false;
+                if (x.Metadata.Gene.IsNullOrEmpty() && y.Metadata.Gene.IsNullOrEmpty()) return x.Sequence.Equals(y.Sequence);
+                if (x.Metadata.Gene.IsNullOrEmpty() != y.Metadata.Gene.IsNullOrEmpty()) return false;
+
+                return x.Metadata.Gene.Equals(y.Metadata.Gene);
+            }
+
+            public override int GetHashCode(IProteinRecord obj)
+            {
+                if (obj.Metadata.Gene == null)
+                    return obj.Sequence.GetHashCode();
+                return obj.Metadata.Gene.GetHashCode();
+            }
+        }
+
+        private Dictionary<IProteinRecord, PeptideAssociationGroup> CalculateProteinOrGeneGroups(MappingResultsInternal results, bool geneLevel, ILongWaitBroker broker)
         {
             var _peptideGroupToProteins = new Dictionary<PeptideAssociationGroup, List<IProteinRecord>>();
 
-            foreach(var kvp in AssociatedProteins)
+            var proteinOrGeneToPeptideGroup = AssociatedProteins;
+            if (geneLevel)
+            {
+                // gene to protein to peptides; the top level dictionary uses the GeneLevelEqualityComparer
+                var geneToPeptides = new Dictionary<IProteinRecord, Dictionary<IProteinRecord, List<PeptideDocNode>>>(new GeneLevelEqualityComparer());
+                foreach (var kvp in AssociatedProteins)
+                    if (!geneToPeptides.ContainsKey(kvp.Key))
+                        geneToPeptides.Add(kvp.Key, new Dictionary<IProteinRecord, List<PeptideDocNode>> { { kvp.Key, kvp.Value.Peptides } });
+                    else
+                        geneToPeptides[kvp.Key][kvp.Key] = kvp.Value.Peptides;
+                // now pick the protein with the longest sequence if it contains all the peptides, or a concatenation of all of the sequences if not
+                IProteinRecord GenerateConcatenatedSequenceIfNecessary(Dictionary<IProteinRecord, List<PeptideDocNode>> proteinToPeptides)
+                {
+                    if (proteinToPeptides.Count == 1)
+                        return proteinToPeptides.Keys.First();
+
+                    var longestProtein = proteinToPeptides.OrderByDescending(kvp2 => kvp2.Key.Sequence.Sequence.Length).First().Key;
+                    var allPeptides = proteinToPeptides.Values.SelectMany(o => o).Distinct().ToList();
+                    if (allPeptides.All(node => longestProtein.Sequence.Sequence.Contains(node.Peptide.Sequence)))
+                        return longestProtein;
+
+                    return new FastaRecord(longestProtein.RecordIndex, 0,
+                        new FastaSequence(longestProtein.Sequence.Name, longestProtein.Sequence.Description, null,
+                            string.Join("", proteinToPeptides.Keys.Select(o => o.Sequence.Sequence))), longestProtein.Metadata);

```

Comment:
@brendanx67 This is what I came up with to pick/make a protein sequence that contains all the member peptides. It will use the longest sequence of the group if it contains all the peptides, and if it doesn't then it'll concatenate all the protein sequences together. Some alternatives might be: concatenate all the peptides instead of the proteins, or only concatenate enough proteins to contain all the peptides (i.e. a parsimonious approach to concatenating).

Or throw this hacky workaround away and go back to the drawing board, making gene group a new FastaSequenceGroup that allows its peptides to map to only some of the parent's FastaSequenceList. I honestly don't know how much work that would be though.